### PR TITLE
Feature: BN 2.0

### PR DESCRIPTION
**Pull request by napalmpapalam**

### CODE REVIEW

File: packages/tools/src/helpers/general.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { HEX_REGEX, NUMBER_REGEX } from '@/const'
+import { RuntimeError } from '@/errors'
+
+export const isHex = (value: string): boolean => HEX_REGEX.test(value)
+
+export const isIntegerString = (value: string): boolean => {
+  return Boolean(value.match(/^(-?)(\d*)$/))
```
Comment (ardier16) [inline packages/tools/src/helpers/general.ts]: '-' is integer string

File: packages/tools/src/helpers/general.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+export const isFixedPointString = (value: string): boolean => {
+  return Boolean(value.match(NUMBER_REGEX))
+}
+
+export function assert(
+  expression: boolean,
+  message: string,
+): asserts expression is true {
+  if (expression) throw new RuntimeError(message)
```
Comment (ardier16) [inline packages/tools/src/helpers/general.ts]: why assert works vice versa comparing to Node assert?
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

File: packages/tools/src/types/bn.ts [hunk 3]
```diff
@@ -1,13 +1,35 @@
...
+  precision: number
+  rounding: BN_ROUNDING
+  format: BnFormatConfig
+}
+
+export type BnFormatConfig = {
+  // string to prepend
+  prefix?: string
+  // decimal separator
+  decimalSeparator?: string
```
Comment (ardier16) [inline packages/tools/src/types/bn.ts]: thx cap 🫡

File: packages/tools/src/const/bn.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { DECIMALS } from '@/enums'
+
+export const DEFAULT_BN_PRECISION = DECIMALS.YOCTO
+export const NUMBER_REGEX = /^(-?)(\d*)\.?(\d*)$/
```
Comment (ardier16) [inline packages/tools/src/const/bn.ts]: '-' and '-.' are valid numbers here

File: packages/tools/src/const/bn.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+export const HEX_REGEX = /^0x[0-9a-f]+$/i
```
Comment (ardier16) [inline packages/tools/src/const/bn.ts:3]: why not uppercased `A-F` too?

File: packages/tools/src/bn/format.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  return [prefix, negative, whole, decimalSeparator, fraction, suffix].join('')
+}
+
+const insertCharEveryN = (value: string, char: string, n: number) => {
+  const result = value.split('').reduce((acc, i, idx) => {
+    if (idx && !(idx % n)) return acc + char + i
+    return acc + i
+  }, '')
+
+  return result.endsWith(char) ? result.slice(0, -1) : result
```
Comment (ardier16) [inline packages/tools/src/bn/format.ts:54]:
```
> insertCharEveryN('berta', 'a', 3)
< 'berat'
```

File: packages/tools/src/bn/format.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  return [prefix, negative, whole, decimalSeparator, fraction, suffix].join('')
+}
+
+const insertCharEveryN = (value: string, char: string, n: number) => {
+  const result = value.split('').reduce((acc, i, idx) => {
+    if (idx && !(idx % n)) return acc + char + i
+    return acc + i
+  }, '')
```
Comment (ardier16) [inline packages/tools/src/bn/format.ts]: may be
```const result = value.replace(new RegExp(`(.{${n}})`, 'g'), `$1${char}`)```

File: packages/tools/src/bn/format.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  /* eslint-enable prefer-const */
+
+  const expression =
+    !Number.isInteger(groupSize) || !Number.isInteger(fractionGroupSize)
+
+  assert(expression, 'groupSize and fractionGroupSize must be an integer')
+
+  let [whole, fraction] = value.split('.')
+
+  const negative = whole.startsWith('-') ? '-' : ''
```
Comment (ardier16) [inline packages/tools/src/bn/format.ts]: maybe `sign` instead?

File: packages/tools/src/bn/format.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  if (groupSize && groupSeparator) {
+    whole = insertCharEveryN(whole, groupSeparator, groupSize)
+  }
+
+  if (decimals === 0) {
+    fraction = ''
+    decimalSeparator = ''
+  }
+
+  if (decimals) {
```
Comment (ardier16) [inline packages/tools/src/bn/format.ts]: duplicate conditional
let's either use ternary or else

File: packages/tools/src/bn/assertions.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  const expression = isGreater
+    ? decimals < currentDecimals
+    : decimals > currentDecimals
+
+  const message = isGreater
+    ? 'Provided decimals cannot be less than the current decimals'
+    : 'Provided decimals cannot be greater than the current decimals'
+
+  assert(expression, message)
```
Comment (ardier16) [inline packages/tools/src/bn/assertions.ts]: IMO `if (op === BN_ASSERT_DECIMALS_OP.GREATER) { assert(...) } else { assert(...) }` or `if return` is more readable here as it does not require to keep in mind the context

File: packages/tools/src/bn/decimals.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import { BN_ASSERT_DECIMALS_OP } from '@/enums'
+
+import { assertDecimals } from './assertions'
+
+export const toDecimals = (
+  val: bigint,
+  decimals: number,
```
Comment (ardier16) [inline packages/tools/src/bn/decimals.ts]: why we pass `decimals` to function `toDecimals`? 🤔

File: packages/tools/src/bn/decimals.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+): bigint => {
+  return decimals > actualDecimals
+    ? toGreaterDecimals(val, decimals, actualDecimals)
+    : toLessDecimals(val, decimals, actualDecimals)
+}
+
+export const toGreaterDecimals = (
+  val: bigint,
+  decimals: number,
+  currentDecimals: number,
```
Comment (ardier16) [inline packages/tools/src/bn/decimals.ts:17]: why it's `actualDecimals` above, but `currentDecimals` here?

File: packages/tools/src/bn/parsers.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import { NUMBER_REGEX } from '@/const'
+import { assert, isFixedPointString } from '@/helpers'
+import type { BnConfig, BnConfigLike } from '@/types'
+
+import { BN } from './bn'
+
+export const parseNumberString = (_value: string): string => {
+  let val = _value.trimStart().trimEnd()
```
Comment (ardier16) [inline packages/tools/src/bn/parsers.ts]: why not just `trim`? 🤔

**Review by ardier16 — CHANGES_REQUESTED** [review]

File: packages/tools/src/const/bn.ts (hunk 5 above)
Comment (napalmpapalam) [reply packages/tools/src/const/bn.ts:3]: i flag there

File: packages/tools/src/helpers/general.ts (hunk 1 above)
Comment (napalmpapalam) [reply packages/tools/src/helpers/general.ts]: fixed

File: packages/tools/src/types/bn.ts (hunk 3 above)
Comment (napalmpapalam) [reply packages/tools/src/types/bn.ts]: removed

File: packages/tools/src/const/bn.ts (hunk 4 above)
Comment (napalmpapalam) [reply packages/tools/src/const/bn.ts]: fixed

File: packages/tools/src/bn/format.ts (hunk 7 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/format.ts]: yea

File: packages/tools/src/bn/format.ts (hunk 8 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/format.ts]: ye

File: packages/tools/src/bn/format.ts (hunk 9 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/format.ts]: fixed

File: packages/tools/src/bn/decimals.ts (hunk 12 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/decimals.ts:17]: fiixed

File: packages/tools/src/bn/parsers.ts (hunk 13 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/parsers.ts]: fixed

File: packages/tools/src/bn/format.ts (hunk 6 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/format.ts:54]: fixed

File: packages/tools/src/helpers/general.ts (hunk 2 above)
Comment (napalmpapalam) [reply packages/tools/src/helpers/general.ts]: fixed

File: packages/tools/src/bn/assertions.ts (hunk 10 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/assertions.ts]: fixed

File: packages/tools/src/bn/decimals.ts (hunk 11 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/decimals.ts]: fixed

File: packages/tools/src/bn/decimals.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { BN_ASSERT_DECIMALS_OP } from '@/enums'
+
+import { assertDecimals } from './assertions'
+
+export const toDecimals = (
+  val: bigint,
+  currentDecimals: number,
```
Comment (ArtSuslov) [inline packages/tools/src/bn/decimals.ts:7]: BigInt argument should be integer, passing float number will throw (everywhere where you format number to big int)

**Review by ardier16 — APPROVED** [review]

File: packages/tools/src/bn/bn.ts [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+    const withPrecision = parsed * getTens(BN.precision - cfg.decimals)
+    return new BN(withPrecision, cfg)
+  }
+
+  /**
+   * @returns A new {@link BN} if `value` argument is valid integer or float value,
+   * otherwise throws {@link RuntimeError}.
+   * @example
+   * ```ts
+   * const oneEth = BN.fromBigInt(1, 18)
```
Comment (ArtSuslov) [inline packages/tools/src/bn/bn.ts]: fromRaw

File: packages/tools/src/bn/bn.ts [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+    const val = String(value)
+    return new BN(BigInt(parseNumberString(val)), parseConfig(decimalsOrConfig))
+  }
+
+  /**
+   *  @returns A minimum {@link BN} value from the `args`.
+   */
+  public static min(...args: BN[]): BN {
+    const min = args.reduce((min, el) => (el.raw < min.raw ? el : min))
+    return new BN(min.raw, min.config)
```
Comment (ArtSuslov) [inline packages/tools/src/bn/bn.ts]: do you actually need to create new BN? min is already an instance of BN

File: packages/tools/src/bn/bn.ts [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+  }
+
+  /**
+   *  @returns A maximum {@link BN} value from the `args`.
+   */
+  public static max(...args: BN[]): BN {
+    const max = args.reduce((min, el) => (el.raw > min.raw ? el : min))
+    return new BN(max.raw, max.config)
```
Comment (ArtSuslov) [inline packages/tools/src/bn/bn.ts]: same

File: packages/tools/src/bn/bn.ts [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+    return this.cmp(other) >= 0
+  }
+
+  /**
+   * @returns A new {@link BN} whose value is the square root of `this`.
+   * @throws {@link RuntimeError} if `BN.precision` is not even number.
+   */
+  public sqrt(): BN {
+    const expression = BN.precision > 1 && BN.precision % 2 === 0
+    assert(expression, 'sqrt requires precision to be even number')
```
Comment (ArtSuslov) [inline packages/tools/src/bn/bn.ts]: discussed

File: packages/tools/src/bn/bn.ts [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+   * @returns A new {@link BN} whose value is absolute `this`.
+   */
+  public abs(): BN {
+    return this.isPositive ? this : this.negated()
+  }
+
+  /**
+   * @returns A new {@link BN} whose value is `this` raised to the power of `exponent`.
+   */
+  public pow(exponent: number): BN {
```
Comment (ArtSuslov) [inline packages/tools/src/bn/bn.ts:310]: float validation

**Review by ArtSuslov — CHANGES_REQUESTED** [review]

File: packages/tools/src/bn/bn.ts (hunk 19 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/bn.ts:310]: fixed

File: packages/tools/src/bn/bn.ts (hunk 17 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/bn.ts]: fixed

File: packages/tools/src/bn/bn.ts (hunk 16 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/bn.ts]: fixed

File: packages/tools/src/bn/bn.ts (hunk 15 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/bn.ts]: fixed

File: packages/tools/src/bn/decimals.ts (hunk 14 above)
Comment (napalmpapalam) [reply packages/tools/src/bn/decimals.ts:7]: fixed

**Review by ArtSuslov — APPROVED** [review]

**Pull request merged**